### ISSUE TITLE
Scaling issue

### ISSUE THREAD

**Issue by rebcabin** [Engelberg/instaparse#164]
I have huge files (18+M) to parse the JVM runs out of heap (even 12G of heap). I suspect this may be out of the use-case parameters for instaparse, but I thought I'd bring it up in case you'd be interested in seeing it and advising me on it. 

I am parsing FBX ASCII files exported from Blender. I have a successful and concise grammar for such files. I slurp the files into memory, then use instaparse to convert the files to ASTs in HICCUP, and then some regular Clojure functions to convert the ASTs to native Clojure data structures like keywords, vectors, sequences, and the like. Using instaparse's "transform" functionality is on my to-do list. 

The issue is that I cannot scale my parser. It will parse an FBX file containing 5,000 lines and 750KB in less than two minutes, and that's fast enough for my applications, but one of my full models has about 142,000 lines and 18M bytes, and JVM runs out of heap memory, even when given 12G. 

I know the native Clojure reader and data structures scale: I verified that I can read 100K vertices / second out of a file of 135MB using the Clojure reader. That's plenty fast enough and scale enough for my applications. But without the parser, I am going to have to write a preprocessor in C (or something else) to convert the FBX file into native Clojure structures. 

Please let me know if this is a suitable and / or interesting issue for you to look at and I will provide details. Because the samples are large, I suspect I will have to work out some way to get them to you other than pasting here.

**Comment by Engelberg** [comment]
Yes, this is outside the scope of what instaparse can do, because it is so memory-intensive.  

Is the fbx file comprised of independent objects?  If so, one option would be to preprocess the file, chopping it into the separate objects, and then run instaparse on the strings for the individual objects.  I created an optional flag for instaparse `:optimize :memory` which attempts to do this chopping up automatically, but it is imperfect.

**Comment by Engelberg** [comment]
I would love to have the example grammar and file, though, because it gives me a concrete test I can use to try to make improvements to the memory usage.  Probably what I'd need to do is have some way to say, "If you parse this particular non-terminal successfully, we're not going to backtrack, so garbage-collect all the corresponding structures for backtracking."

**Comment by rebcabin** [comment]
I will pack up something for you soon. Glad to hear you find the problem interesting!

**Comment by rebcabin** [comment]
I managed to write an equivalent parser in yacc-lex that translates an FBX into Clojure very quickly (18M FBX in a second or two). I wrote an "attolisp" [sic, 1000x smaller than femtolisp] in the yacc code to build the tree in memory, then spit it out in Clojure format. Clojure reads the model in a couple of seconds, too.  Pprint takes a couple of minutes, but processes the 17M model without difficulty. If I make the embedded lisp any smaller, I still have "zeptolisp" and "yoctolisp" available as names. I had first tried to hand-write a parser in Statecharts with Yakindu, but that was getting difficult. I can resurrect the statecharts form by hand-transcribing the state machine that yacc produces back into Yakindu. That might be an amusing exercise, but not important for getting the job done.
